ansible/files/admin_api_scripts/pg_upgrade_initiate.sh

```diff
@@ -37,78 +69,143 @@ cleanup() {
+    for EXTENSION in "${EXTENSIONS_TO_DISABLE[@]}"; do
+        EXTENSION_ENABLED=$(run_sql -A -t -c "SELECT EXISTS(SELECT 1 FROM pg_extension WHERE extname = '${EXTENSION}');")
+        if [ "$EXTENSION_ENABLED" = "t" ]; then
+            echo "Disabling extension ${EXTENSION}"
+            run_sql -c "DROP EXTENSION IF EXISTS ${EXTENSION} CASCADE;"
+            cat << EOF >> $POST_UPGRADE_EXTENSION_SCRIPT
+DO \$\$
+BEGIN
+    IF EXISTS (SELECT 1 FROM pg_available_extensions WHERE name = '${EXTENSION}') THEN
+        CREATE EXTENSION IF NOT EXISTS ${EXTENSION} CASCADE;
+    END IF;
+END;
+\$\$;
+EOF
+        fi
+    done
+}
 
+function initiate_upgrade {
     mkdir -p "$MOUNT_POINT"
-    mount "$BLOCK_DEVICE" "$MOUNT_POINT"
-    resize2fs "$BLOCK_DEVICE"
+    SHARED_PRELOAD_LIBRARIES=$(cat /etc/postgresql/postgresql.conf | grep shared_preload_libraries | sed "s/shared_preload_libraries = '\(.*\)'.*/\1/")
 
-    SHARED_PRELOAD_LIBRARIES=$(cat /etc/postgresql/postgresql.conf | grep shared_preload_libraries |  sed "s/shared_preload_libraries = '\(.*\)'.*/\1/")
-    PGDATAOLD=$(cat /etc/postgresql/postgresql.conf | grep data_directory | sed "s/data_directory = '\(.*\)'.*/\1/")    
+    # Wrappers officially launched in PG15; PG14 version is incompatible
+    if [[ "$OLD_PGVERSION" =~ 14* ]]; then
+        SHARED_PRELOAD_LIBRARIES=$(echo $SHARED_PRELOAD_LIBRARIES | sed "s/wrappers, //")
+    fi
+
+    PGDATAOLD=$(cat /etc/postgresql/postgresql.conf | grep data_directory | sed "s/data_directory = '\(.*\)'.*/\1/")
 
     PGDATANEW="$MOUNT_POINT/pgdata"
-    PGBINNEW="/tmp/pg_upgrade_bin/$PGVERSION/bin"
-    PGSHARENEW="/tmp/pg_upgrade_bin/$PGVERSION/share"
+    PG_UPGRADE_BIN_DIR="/tmp/pg_upgrade_bin/$PGVERSION"
+    PGBINNEW="$PG_UPGRADE_BIN_DIR/bin"
+    PGLIBNEW="$PG_UPGRADE_BIN_DIR/lib"
+    PGSHARENEW="$PG_UPGRADE_BIN_DIR/share"
 
+    # running upgrade using at least 1 cpu core
+    WORKERS=$(nproc | awk '{ print ($1 == 1 ? 1 : $1 - 1) }')
+    
+    echo "1. Extracting pg_upgrade binaries"
     mkdir -p "/tmp/pg_upgrade_bin"
-    tar zxvf "/tmp/persistent/pg_upgrade_bin.tar.gz" -C "/tmp/pg_upgrade_bin"
+    tar zxf "/tmp/persistent/pg_upgrade_bin.tar.gz" -C "/tmp/pg_upgrade_bin"
 
     # copy upgrade-specific pgsodium_getkey script into the share dir
+    chmod +x "/root/pg_upgrade_pgsodium_getkey.sh"
     cp /root/pg_upgrade_pgsodium_getkey.sh "$PGSHARENEW/extension/pgsodium_getkey"
-    chmod +x "$PGSHARENEW/extension/pgsodium_getkey"
+    if [ -d "/var/lib/postgresql/extension/" ]; then
+        cp /root/pg_upgrade_pgsodium_getkey.sh "/var/lib/postgresql/extension/pgsodium_getkey"
+        chown postgres:postgres "/var/lib/postgresql/extension/pgsodium_getkey"
+    fi
+
+    chown -R postgres:postgres "/tmp/pg_upgrade_bin/$PGVERSION"
+
+    # Make latest libpq available to pg_upgrade
+    mkdir -p /usr/lib/postgresql/lib/aarch64
+    if [ ! -L /usr/lib/postgresql/lib/aarch64/libpq.so.5 ]; then
```

Comment:
hmm do we block upgrades on the older x86 instances at some early stage?

Reply:
Not an explicit one - it's tied down to version eligibility since the last version supporting x86 was 12.5.
Will add additional checks in the admin cli and upcoming API support